src/nccl_ofi_net.c

```diff
@@ -1009,6 +1009,61 @@ int nccl_net_ofi_domain_fini(nccl_net_ofi_domain_t *domain)
 }
 
 
+int nccl_net_ofi_domain_release_all(nccl_net_ofi_device_t *device)
+{
+	int ret, first_error = 0, domain_num;
+
+	assert(device != NULL);
+	nccl_net_ofi_domain_t *domain, *domain_tmp;
+	nccl_net_ofi_ep_t *ep, *ep_tmp;
+
+	domain_num = HASH_COUNT(device->domain_table);
+	assert(domain_num > 0);
+	HASH_ITER(hh, device->domain_table, domain, domain_tmp) {
+		/* For each domain, clean up its endpoints. */
+		nccl_net_ofi_mutex_lock(&domain->domain_lock);
```

Comment:
This isn't very abstraction friendly.  This should be a call into the domain class that handles all this (including taking the lock).